src/lib/y2storage/blk_device.rb

```diff
@@ -366,6 +383,20 @@ def multipath
       children.find { |dev| dev.is?(:multipath) }
     end
 
+    # Bcache device defined on top of the device
+    #
+    # @return [Bcache, nil] nil if the device is not part of any bcache
+    def bcache
+      children.find { |dev| dev.is?(:bcache) }
+    end
+
+    # Bcache caching set device defined on top of the device
+    #
+    # @return [BcacheCset, nil] nil if the device is not used as bcache caching set
+    def direct_bcache_cset
```

Comment:
I find the usage of `direct_` to be inconsistent between `#bcache` and `#direct_bcache_set`.

Can a bcache be defined with a LUKS as backing device? In that case, I would expect two methods:

* `#direct_bcache`
  * For `sda1 -> LUKS -> bcache0` it returns nil
  * For `sda1 -> bcache0` it returns bcache0

* `#bcache`
  * For `sda1 -> LUKS -> bcache0` it returns bcache0
  * For `sda1 -> bcache0` it returns bcache0

And the other way around. If a LUKS device can be a caching device, I would expect to have both `#direct_bcache_set` and `#bcache_set`.

Moreover, I find the usage of terms like "on top" potentially confusing in the descriptions, specially since it's used for both the bcache and the cset devices. What about something like "Bcache defined on top of the device, i.e. a Bcache device that uses this one as backing device".

Reply:
Well, to be honest I do not play with LUKS yet on top of it. But I worry using luks on top of caching device kills performance.

Reply:
Is not about performance or how correct it is, it's about using method names in a consistent way. I don't see the point in having a `#bcache` method that returns bcache devices placed directly (no LUKS in the middle) "on top" of this device and then a `#direct_bcache_cset` method that returns csets also placed directly on the device. It's inconsistent to each other and to the other methods in the class.

Reply:
hmm, that scares me a bit that there is so many new methods needed. Would be possible to have only non-direct one and if needed define direct as something that takes a block?

like
```
def bcache
  children.find { |dev| dev.is?(:bcache) }
end

def direct
  encrypted? ? <NullObject returning nil> : self
end

# and then usage
dev.direct.bcache
dev.bcache
```

because otherwise we need for any new device device both methods, which looks quite bad for me as API for blk_device is already quite big.

Reply:
btw this pull request also shows how many place I forgot in initial support for bcache. If there is new type of device it needs beside wrapper also add it to blk_device, devicegraph and also everywhere for downcast. I am not sure where it will lead with "read-only" LVMCache and other new potential types. API for some "god" objects is already quite big ( blk_device and devicegraph ).